ecommerce/models.py

```diff
@@ -154,3 +156,97 @@ def save(self, *args, **kwargs):
     def __str__(self):
         """Description for CoursePrice"""
         return "CoursePrice for {}, price={}, is_valid={}".format(self.course_run, self.price, self.is_valid)
+
+
+class Coupon(Model):
+    """
+    Model for a coupon. This stores the discount for the coupon, how many people can use it, and how many times
+    a person can use it.
+
+    When a coupon is redeemed by a purchaser the counter on this object is decremented
+    and a UserCoupon object is created for that particular purchaser.
+    """
+    PERCENT_DISCOUNT = 'percent-discount'
+    FIXED_DISCOUNT = 'fixed-discount'
+
+    AMOUNT_TYPES = [PERCENT_DISCOUNT, FIXED_DISCOUNT]
+
+    COURSE_RUN = 'course-run'
+    COURSE = 'course'
+    PROGRAM = 'program'
+    PRODUCT_TYPES = [PROGRAM, COURSE, COURSE_RUN]
+
+    # The coupon code used for redemption by the purchaser in the user interface.
+    # If blank, the purchaser may not redeem this coupon through the user interface,
+    # though it may be redeemed in their name by an administrator.
+    coupon_code = TextField(null=True, blank=True)
+    product_type = CharField(
+        choices=[(_type, _type) for _type in PRODUCT_TYPES],
+        max_length=30,
+    )
+    product_id = IntegerField(null=False)
+    # percent or fixed discount
+    amount_type = CharField(
+        choices=[(_type, _type) for _type in AMOUNT_TYPES],
+        max_length=30,
+    )
+    # Either a number from 0 to 1 representing a percent, or a fixed value for discount
+    amount = DecimalField(decimal_places=2, max_digits=20)
+
+    # Number of people this coupon can be redeemed by
+    num_coupons_available = IntegerField(null=False)
+    # Number of times a person can redeem a coupon
+    num_redemptions = IntegerField(null=False)
```

Comment:
We will need to do this at the point where the user is requesting to redeem the coupon. I'm thinking of something like this pseudocode:

    def redeem_coupon(coupon, user):
        with transaction.atomic():
            try:
                user_coupon = UserCoupon.objects.get(user=user, coupon=coupon)
            except UserCoupon.DoesNotExist:
                if coupon.num_coupons_available == 0:
                    raise ValidationError("No coupons left")
                    
                coupon.num_coupons_available -= 1
                coupon.save()
                user_coupon = UserCoupon.objects.create(user=user, coupon=coupon)
            
            if user_coupon.available_redemptions == 0:
                raise ValidationError("Coupon has no more redemptions left")
            user_coupon.available_redemptions -= 1
            user_coupon.save()

        order = create_unfulfilled_order()
        RedeemedCoupon.objects.create(order=order, coupon=coupon)
        order.status = 'fulfilled'
        order.save()